File: klinara-ios/klinara-ios/Services/Booking/BookingService.swift
import Foundation

/// Takvim çekirdeğinin uçları — `apps/api/src/modules/booking`.
///
/// Şube kapsamı iki farklı şekilde geçer ve karıştırılmamalıdır: `availability`
/// ve `calendar/*` uçları `@RequireBranchScope`'tur (`X-Branch-Id` başlığı
/// ``APIClient`` tarafından her istekte gönderilir), randevu uçları ise şubeyi
/// **gövdede/sorguda** alır.
protocol BookingService: Sendable {

    /// `GET /availability` — verilen hizmet dizisi için uygun slotlar.
    func availability(_ query: AvailabilityQuery) async throws -> AvailabilityResponse

    /// `GET /appointments` — filtreli, cursor sayfalamalı liste.
    func appointments(_ query: AppointmentListQuery) async throws -> Page<CalendarEntry>

    /// `GET /appointments/:id`
    func appointment(id: String) async throws -> Appointment

    /// `GET /appointments/:id/history`
    func history(id: String) async throws -> [AppointmentHistoryEntry]

    /// `POST /appointments` — `Idempotency-Key` ile tekrar güvenli.
    func create(
        _ input: CreateAppointmentInput,
        idempotencyKey: String
    ) async throws -> Appointment

    /// `PATCH /appointments/:id` — yalnız notu değiştirir, `If-Match` zorunlu.
    ///
    /// `notes` opsiyonel değil: `nil` göndermek notu **siler**, sunucudaki
    /// davranışın aynısı. Mevcut notu korumak isteyen onu geri göndermeli.
    func updateNotes(id: String, version: Int, notes: String?) async throws -> Appointment

    /// `POST /appointments/:id/reschedule` — `If-Match` zorunlu.
    func reschedule(
        id: String,
        version: Int,
        _ input: RescheduleAppointmentInput
    ) async throws -> Appointment

    /// `POST /appointments/:id/cancel` — `If-Match` **almaz**, sürümü sunucu okur.
    func cancel(id: String, reason: String?) async throws -> Appointment

    /// `POST /appointments/:id/status`
    func changeStatus(id: String, _ input: ChangeAppointmentStatusInput) async throws -> Appointment

    /// `GET /calendar/day` — `X-Branch-Id` zorunlu.
    func calendarDay(_ query: CalendarDayQuery) async throws -> CalendarResponse

    /// `GET /calendar/week` — `X-Branch-Id` zorunlu.
    func calendarWeek(_ query: CalendarWeekQuery) async throws -> CalendarResponse

    /// `GET /calendar/staff` — `X-Branch-Id` zorunlu.
    func calendarStaff(_ query: CalendarStaffQuery) async throws -> CalendarResponse
}

struct LiveBookingService: BookingService {

    private let client: APIClient

    init(client: APIClient) {
        self.client = client
    }

    // MARK: Uygunluk

    func availability(_ query: AvailabilityQuery) async throws -> AvailabilityResponse {
        var items = [
            URLQueryItem(name: "branchId", value: query.branchId),
            // Sunucu hem `a,b` hem tekrarlı parametreyi kabul ediyor; virgüllü
            // biçim URL'i kısa tutar ve sırayı gözle doğrulanabilir kılar.
            URLQueryItem(name: "serviceIds", value: query.serviceIds.joined(separator: ",")),
            URLQueryItem(name: "from", value: KlinaraCoding.timestamp(query.from)),
            URLQueryItem(name: "to", value: KlinaraCoding.timestamp(query.to)),
        ]
        if let staffProfileId = query.staffProfileId {
            items.append(URLQueryItem(name: "staffProfileId", value: staffProfileId))
        }
        return try await client.send(APIRequest.get("availability", query: items))
    }

    // MARK: Randevu

    func appointments(_ query: AppointmentListQuery) async throws -> Page<CalendarEntry> {
        var items = [
            URLQueryItem(name: "from", value: KlinaraCoding.timestamp(query.from)),
            URLQueryItem(name: "to", value: KlinaraCoding.timestamp(query.to)),
        ]
        if let branchId = query.branchId {
            items.append(URLQueryItem(name: "branchId", value: branchId))
        }
        if let customerId = query.customerId {
            items.append(URLQueryItem(name: "customerId", value: customerId))
        }
        if let staffProfileId = query.staffProfileId {
            items.append(URLQueryItem(name: "staffProfileId", value: staffProfileId))
        }
        if !query.status.isEmpty {
            let raw = query.status.map(\.rawValue).joined(separator: ",")
            items.append(URLQueryItem(name: "status", value: raw))
        }
        if let limit = query.limit {
            items.append(URLQueryItem(name: "limit", value: String(limit)))
        }
        if let cursor = query.cursor {
            items.append(URLQueryItem(name: "cursor", value: cursor))
        }
        return try await client.send(APIRequest.get("appointments", query: items))
    }

    func appointment(id: String) async throws -> Appointment {
        try await client.send(APIRequest.get("appointments/\(id)"))
    }

    func history(id: String) async throws -> [AppointmentHistoryEntry] {
        let response: ListEnvelope<AppointmentHistoryEntry> = try await client.send(
            APIRequest.get("appointments/\(id)/history")
        )
        return response.data
    }

    func create(
        _ input: CreateAppointmentInput,
        idempotencyKey: String
    ) async throws -> Appointment {
        try await client.send(
            APIRequest.post("appointments", body: input, idempotencyKey: idempotencyKey)
        )
    }

    func updateNotes(id: String, version: Int, notes: String?) async throws -> Appointment {
        try await client.send(
            APIRequest.patch(
                "appointments/\(id)",
                body: UpdateAppointmentInput(notes: notes),
                ifMatch: weakETag(version)
            )
        )
    }

    func reschedule(
        id: String,
        version: Int,
        _ input: RescheduleAppointmentInput
    ) async throws -> Appointment {
        try await client.send(
            APIRequest.post(
                "appointments/\(id)/reschedule",
                body: input,
                ifMatch: weakETag(version)
            )
        )
    }

    func cancel(id: String, reason: String?) async throws -> Appointment {
        try await client.send(
            APIRequest.post("appointments/\(id)/cancel", body: CancelAppointmentInput(reason: reason))
        )
    }

    func changeStatus(id: String, _ input: ChangeAppointmentStatusInput) async throws -> Appointment {
        try await client.send(APIRequest.post("appointments/\(id)/status", body: input))
    }

    // MARK: Takvim

    func calendarDay(_ query: CalendarDayQuery) async throws -> CalendarResponse {
        var items = [
            URLQueryItem(name: "branchId", value: query.branchId),
            URLQueryItem(name: "date", value: query.date),
        ]
        if let staffProfileId = query.staffProfileId {
            items.append(URLQueryItem(name: "staffProfileId", value: staffProfileId))
        }
        return try await client.send(APIRequest.get("calendar/day", query: items))
    }

    func calendarWeek(_ query: CalendarWeekQuery) async throws -> CalendarResponse {
        var items = [
            URLQueryItem(name: "branchId", value: query.branchId),
            URLQueryItem(name: "weekStart", value: query.weekStart),
        ]
        if let staffProfileId = query.staffProfileId {
            items.append(URLQueryItem(name: "staffProfileId", value: staffProfileId))
        }
        return try await client.send(APIRequest.get("calendar/week", query: items))
    }

    func calendarStaff(_ query: CalendarStaffQuery) async throws -> CalendarResponse {
        let items = [
            URLQueryItem(name: "branchId", value: query.branchId),
            URLQueryItem(name: "staffProfileId", value: query.staffProfileId),
            URLQueryItem(name: "from", value: KlinaraCoding.timestamp(query.from)),
            URLQueryItem(name: "to", value: KlinaraCoding.timestamp(query.to)),
        ]
        return try await client.send(APIRequest.get("calendar/staff", query: items))
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/services/booking/BookingService.kt
package com.klinara.android.services.booking

import com.klinara.android.services.networking.Page

/**
 * Takvim ve randevu uçları. Uç yolları doc yorumlarında; `API_DEVELOPMENT.md` Faz 3.
 *
 * A3.1'de üç okuma metodu var; arayüz A3.3 (yaşam döngüsü) ve A3.4 (oluşturma) ile
 * büyür. Boş imza tahmini kodlanmadı: okunmamış bir ucun imzası her batch'te
 * "refactor" edilir ve bu, ilerleme gibi görünen bir çalkantıdır (A0.5 kararı).
 *
 * **Şube iki yerden gider.** `calendar/day` ve `calendar/week` hem `?branchId=` sorgu parametresi
 * hem `X-Branch-Id` başlığı ister (`@RequireBranchScope`); başlığı `HeaderInterceptor`
 * koyuyor, sorgu parametresi burada. İkisinden biri eksikse 400 `VALIDATION_FAILED`.
 */
interface BookingService {
    /** `GET calendar/day` — tek günün randevuları + yoğunluk kovaları. */
    suspend fun calendarDay(query: CalendarDayQuery): CalendarResponse

    /** `GET calendar/week` — yedi günün randevuları, tek istekte. */
    suspend fun calendarWeek(query: CalendarWeekQuery): CalendarResponse

    /**
     * `GET appointments` — tarih aralığına göre imleçli liste.
     *
     * Takvim ekranı bunu kullanmaz (`calendar/day` daha dar ve yoğunluk da getiriyor);
     * müşteri kartının randevu geçmişi (A4) ve aralık sorguları için var.
     */
    suspend fun appointments(query: AppointmentListQuery): Page<CalendarEntry>

    // --- Yaşam döngüsü (A3.3) ---

    /**
     * `GET appointments/:id`
     *
     * Takvim satırı yetmez: `origin`, `cancellationReason` ve tampon süreleri yalnız
     * burada, üstelik taze bir `version` gerekiyor.
     *
     * **`practitioner` başkasının randevusunda 404 alır, 403 değil** — sunucu bunu
     * bilerek yapıyor (403 kaydın VAR OLDUĞUNU sızdırırdı). Ekran "yetkiniz yok"
     * dememeli, "bulunamadı" demeli.
     */
    suspend fun appointment(id: String): Appointment

    /** `GET appointments/:id/history` — salt okunur denetim izi. */
    suspend fun history(id: String): List<AppointmentHistoryEntry>

    /**
     * `PATCH appointments/:id` — **`If-Match` zorunlu**.
     *
     * Sunucuda düzenlenebilir tek alan not. [notes] `null` gönderilirse not SİLİNİR;
     * ekran bunu kullanıcıya söylemeli.
     */
    suspend fun updateNotes(
        id: String,
        version: Int,
        notes: String?,
    ): Appointment

    /**
     * `POST appointments/:id/cancel` — `If-Match` **istemez**.
     *
     * Yanıt `ETag` başlığı da döndürmüyor ama gövde `version` taşıyor; sürüm oradan
     * alınır ve yeniden `GET` gerekmez.
     */
    suspend fun cancel(
        id: String,
        reason: String?,
    ): Appointment

    /**
     * `POST appointments/:id/status`
     *
     * Aynı duruma geçiş bir hata değil, **200 dönen bir no-op**. Geçersiz geçiş
     * 409 `INVALID_STATUS_TRANSITION`; izin eksikse 403 (`appointment:reopen`).
     */
    suspend fun changeStatus(
        id: String,
        status: AppointmentStatus,
        reason: String? = null,
    ): Appointment

    // --- Oluşturma ve erteleme (A3.4) ---

    /** `GET availability` — verilen hizmet dizisi için uygun başlangıçlar. */
    suspend fun availability(query: AvailabilityQuery): AvailabilityResponse

    /**
     * `POST appointments`
     *
     * [idempotencyKey] **her denemede YENİ** olmalı: düzeltilmiş bir gövdeyi aynı
     * anahtarla göndermek 409 `IDEMPOTENCY_CONFLICT` verir. Anahtar ağ tekrarına karşı,
     * çift dokunuşa karşı değil — onu `isSaving` bayrağı engelliyor.
     *
     * Çakışmada 409 `SLOT_CONFLICT` gelir ve gövde **alternatif saatler** taşır; bu bir
     * hata değil, bir bilgidir.
     */
    suspend fun create(
        input: CreateAppointmentInput,
        idempotencyKey: String,
    ): Appointment

    /** `POST appointments/:id/reschedule` — **`If-Match` zorunlu**. */
    suspend fun reschedule(
        id: String,
        version: Int,
        input: RescheduleAppointmentInput,
    ): Appointment
}
